src/java/org/apache/cassandra/streaming/StreamSession.java

```diff
@@ -1338,18 +1339,26 @@ public String toString()
                '}';
     }
 
-    public StringBuilder boundStacktrace(StackTraceElement[] stacktrace, int limit)
+    public static void boundStackTrace(Throwable e, int limit, HashSet<Throwable> visited)
     {
-        if (limit >= stacktrace.length) {
-            throw new IllegalArgumentException("Invalid limit size: " + limit);
-        }
+        if (e == null)
+            return;
+
+        if (visited.contains(e))
+            throw new IllegalArgumentException("Exception cycle detected");
+        else
+            visited.add(e);
```

Comment:
nit: `add` returns boolean saying if it was added or not, so you can do `if (!visited.add(e)) throw new IllegalArgumentException("Exception cycle detected");`